### DISCUSSION TITLE
Implement a programmatic cancel of an updpate

### DISCUSSION THREAD

**Discussion by gaidarjimax** [sparkle-project/Sparkle#2502] [Q&A]
Hello! I am using Sparkle, more specific the SPUUpdater, to update another application. I would like to have an option to cancel an update programmatically when automaticallyDownloadsUpdates is set to true, and SPUAutomaticUpdateDriver is performing the update.

## Possible Fix

The SPUAutomaticUpdateDriver as other SPUUpdateDrivers has abortUpdateWithError method which does exactly what I need, but this method is not reachable from the SPUUpdater interface, or from UserDriver, or the UpdaterDelegate that are passed to the SPUUpdater.

Thank you!

**Comment by zorgiepoo** [comment]
If the user checks for updates they have the option to skip and cancel the update that is staged to be installed (from the user driver).

If you don't check for updates, you won't know if there's an update staged to be installed. The app itself could download and stage the update to be installed automatically, and your updater may be able to pick that up and resume installing it.

That also means there are cases where you do not want to programmatically/automatically cancel an update because the update could have been started by another updater. So the design leans towards that users should cancel updates..

It may be helpful to elaborate more on your use case / scenario and why you want to cancel an update that has already been downloaded.

**Reply by gaidarjimax** [reply]
Thank you for the quick reply. 
I am trying to avoid showing any UI that Sparkle shows. I perform checks for update using checkForUpdatesInformation, and perform the update itself using checkForUpdateInBackground, having automaticallyDownloadsUpdates set to true. I would like to be able to interrupt(abort) any activity initiated by checkForUpdateInBackground, because the user might accidentally click on the update button from my UI. If the user opts to interrupt(abort) the checkForUpdateInBackground activity, it should stop everything that it was doing. As a solution I have two options - kill the process that called the checkForUpdateInBackground(which calls the abortUdpate of SPUAutomaticUpdateDriver underneath) and second option to call the abortUpdateWithError of SPUAutomaticUpdateDriver manually, and the problem is that abortUpdateWithError is not reachable without changing the Sparkle source code.
May be there is a way of doing it gracefully that I am overlooking, or a reason why I should not interrupt(abort) updates(except the case when another updater is performing the same update)?
Thank you!

**Reply by zorgiepoo** [reply]
These are some constraints / normalities you should know:

* Using checkForUpdatesInformation and then checkForUpdateInBackground incurs overhead of checking for the update twice and is not a recommended approach
* You should not set automaticallyDownloadsUpdates or other updater properties automatically that persist for the bundle without user interaction (and if you do set it, not set it all the time, check the API docs). They are only meant for the user themselves changing settings. Otherwise you will be making changes for the updater settings of the app that will affect the app checking for updates too.
* Apps that use Sparkle do not show update progress when updates are automatically being downloaded in the background. Instead there is a short period of time where an update is not checkable and UI is disabled. There are other short periods of time where an update is not checkable. You need to use [canCheckForUpdates](https://sparkle-project.github.io/documentation/api-reference/Classes/SPUUpdater.html#/c:objc(cs)SPUUpdater(py)canCheckForUpdates) and disable appropriate UI. The user driver is meant for showing the UI of the update progress. If the update has finished downloading, Sparkle will allow checking for updates again and the user can resume that update.

Stopping an ongoing download when the user wants to update an app is also not a good approach.
